File: implementations/lua-jsonschema/bowtie_jsonschema.lua
local jsonschema = require 'jsonschema'
local json = require 'json'

STARTED = false

io.stdout:setvbuf 'no'

local handle = io.popen 'luarocks show jsonschema --mversion'
local jsonschema_version = handle:read '*a'
handle:close()

local function resolve_in_registry(registry)
  return function(url)
    return registry[url]
  end
end

local cmds = {
  start = function(request)
    assert(request.version == 1, 'Wrong version!')
    STARTED = true
    local os_platform_handle = io.popen 'uname'
    local os_platform = os_platform_handle:read '*l'
    os_platform_handle:close()

    local os_version_handle = io.popen 'uname -r'
    local os_version = os_version_handle:read '*l'
    os_version_handle:close()

    local lua_version = (function()
      local temp = {}
      for str in _VERSION:gmatch '%S+' do
        table.insert(temp, str)
      end
      return temp[2]
    end)()

    return {
      version = 1,
      implementation = {
        language = 'lua',
        name = 'jsonschema',
        version = jsonschema_version,
        homepage = 'https://github.com/api7/jsonschema',
        issues = 'https://github.com/api7/jsonschema/issues',
        source = 'https://github.com/api7/jsonschema',

        dialects = {
          'http://json-schema.org/draft-07/schema#',
          'http://json-schema.org/draft-06/schema#',
          'http://json-schema.org/draft-04/schema#',
        },
        os = os_platform,
        os_version = os_version,
        language_version = lua_version,
      },
    }
  end,
  dialect = function(_)
    assert(STARTED, 'Not started!')
    return { ok = false }
  end,
  run = function(request)
    assert(STARTED, 'Not started!')

    local registry = request.case.registry
    local opts = { external_resolver = resolve_in_registry(registry) }
    local ok, result = xpcall(jsonschema.generate_validator, debug.traceback, request.case.schema, opts)
    if not ok then
      return {
        seq = request.seq,
        errored = true,
        context = { traceback = result },
      }
    end
    local results = {}
    for _, test in ipairs(request.case.tests) do
      table.insert(results, { valid = result(test.instance) })
    end
    return { seq = request.seq, results = results }
  end,
  stop = function(_)
    assert(STARTED, 'Not started!')
    os.exit(0)
  end,
}

local request, response
for line in io.lines() do
  request = json.decode(line)
  response = cmds[request.cmd](request)
  io.write(json.encode(response) .. '\n')
end

File: implementations/js-json-schema/bowtie_json_schema.js
import os from "os";
import process from "process";
import readline from "readline/promises";

import { Validator } from "@cfworker/json-schema";
import packageJson from "./node_modules/@cfworker/json-schema/package.json" with { type: "json" };

const json_schema_version = packageJson.version;

const DRAFTS = {
  "https://json-schema.org/draft/2020-12/schema": "2020-12",
  "https://json-schema.org/draft/2019-09/schema": "2019-09",
  "http://json-schema.org/draft-07/schema#": "7",
  "http://json-schema.org/draft-04/schema#": "4",
};

const stdio = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  terminal: false,
});

function send(data) {
  console.log(JSON.stringify(data));
}

var started = false;
var dialect = null;

const cmds = {
  start: async (args) => {
    console.assert(args.version === 1, { args });
    started = true;
    return {
      version: 1,
      implementation: {
        language: "javascript",
        name: "cfworker-json-schema",
        version: json_schema_version,
        homepage:
          "https://github.com/cfworker/cfworker/blob/main/packages/json-schema/README.md",
        issues: "https://github.com/cfworker/cfworker/issues",
        source: "https://github.com/cfworker/cfworker",

        dialects: [
          "https://json-schema.org/draft/2020-12/schema",
          "https://json-schema.org/draft/2019-09/schema",
          "http://json-schema.org/draft-07/schema#",
          "http://json-schema.org/draft-04/schema#",
        ],
        os: os.platform(),
        os_version: os.release(),
        language_version: process.version,
      },
    };
  },

  dialect: async (args) => {
    console.assert(started, "Not started!");
    dialect = args.dialect;
    return { ok: true };
  },

  run: async (args) => {
    console.assert(started, "Not started!");

    const testCase = args.case;

    try {
      const validator = new Validator(testCase.schema, DRAFTS[dialect]);
      for (const id in testCase.registry) {
        validator.addSchema(testCase.registry[id], id);
      }

      const results = testCase.tests.map((test) => {
        try {
          const result = validator.validate(test.instance);
          return { valid: result.valid };
        } catch (error) {
          return {
            errored: true,
            context: {
              traceback: error.stack,
              message: error.message,
            },
          };
        }
      });

      return { seq: args.seq, results: results };
    } catch (error) {
      return {
        errored: true,
        seq: args.seq,
        context: {
          traceback: error.stack,
          message: error.message,
        },
      };
    }
  },

  stop: async (_) => {
    console.assert(started, "Not started!");
    process.exit(0);
  },
};

async function main() {
  for await (const line of stdio) {
    const request = JSON.parse(line);
    const response = await cmds[request.cmd](request);
    send(response);
  }
}

main();
